### PR TITLE
feat(difs): download dSYMs from app store connect task

### PR DESCRIPTION
**Pull request by flub**
This adds a new task which can download dSYMs from Apple App Store
Connect and make them available in the project's debug information
files.

### CODE REVIEW

**Comment by flub** [comment]
I'm not going to keep tracking the migrations until i hit the merge button...

@tonyo is there a trick to locally testing celery tasks?  I don't see anything getting spawned and so far didn't manage to figure that out.

File: src/sentry/lang/native/appconnect.py
Comment (Swatinem) [inline src/sentry/lang/native/appconnect.py:168]:
```suggestion
        This config is normally stored as a symbol source of type ``appStoreConnect`` in a
```

File: src/sentry/lang/native/appconnect.py
Comment (Swatinem) [inline src/sentry/lang/native/appconnect.py:210]: should this call a different version other than `pre_release`?

**Review by Swatinem — APPROVED** [review]
approving to unblock, but I had a few questions

File: src/sentry/lang/native/appconnect.py
Comment (flub) [reply src/sentry/lang/native/appconnect.py:61]: Yes, too many ways to name these things.  Fixed.

File: src/sentry/lang/native/appconnect.py
Comment (flub) [reply src/sentry/lang/native/appconnect.py]: oops, good catch

File: src/sentry/tasks/app_store_connect.py
Comment (flub) [reply src/sentry/tasks/app_store_connect.py:85]: hey, at least it's commented out!

**Review by tonyo — CHANGES_REQUESTED** [review]
A few general questions:
* Are we going to run any other tasks on the new task queue?
* What's the expected traffic for the new queue? Like, will it be rising gradually/on-demand, or e.g. we're going to spawn a task for every existing project a few times a day?

**Comment by Swatinem** [comment]
This popped into my head when we did the planning: Do we have pagination for these things? Or do we get **all** the releases all the time? Can we add pagination, and maybe take advantage of the sort order (does it have one?) to exit early?

File: src/sentry/migrations/0210_appconnectbuilds.py
Comment (markstory) [inline src/sentry/migrations/0210_appconnectbuilds.py:59]: We generally also use the `db_table` option to ensure that we get table names that match the rest of the database. This one would be `sentry_appconnectbuild`.

File: src/sentry/models/appconnectbuilds.py
Comment (markstory) [inline src/sentry/models/appconnectbuilds.py:50]: You should also include the `Meta` inner class that defines the `app_label` and `db_table` attributes.

**Comment by flub** [comment]
> This popped into my head when we did the planning: Do we have pagination for these things? Or do we get **all** the releases all the time? Can we add pagination, and maybe take advantage of the sort order (does it have one?) to exit early?

We get all the pages all the time.  This is one of the optimisations around handling API rate limits better, which is listed somewhere as a concern to be resolved (but not this PR)

File: src/sentry/migrations/0210_appconnectbuilds.py
Comment (flub) [reply src/sentry/migrations/0210_appconnectbuilds.py]: thanks, done

File: src/sentry/models/debugfile.py
Comment (untitaker) [inline src/sentry/models/debugfile.py:403]: would never use fstrings or any generic string interpolation with logging, there's value in having more structured log records where the message is separate from parameters

File: src/sentry/lang/native/appconnect.py
Comment (untitaker) [inline src/sentry/lang/native/appconnect.py:166]: I've seen `Config` type pop up a few times. Can we make this a type alias such that we at least know while reading that this is the same config? If the definition is consolidated it makes it easier to fill in with `TypedDict` at a later point too

File: src/sentry/lang/native/appconnect.py
Comment (untitaker) [inline src/sentry/lang/native/appconnect.py:118]: you can use pathlib, not sure if we use it anywhere else though

File: src/sentry/lang/native/appconnect.py
Comment (untitaker) [inline src/sentry/lang/native/appconnect.py:88]: it might be better to return a more specific type here as we probably assume the data is in a valid format (we do the same in the frontend: assume specific types for API responses without validating them, as it's a "trusted" source). Is this `Config` again?

**Review by untitaker — COMMENTED** [review]
reviewing for typing, can't speak for migration though

File: src/sentry/models/appconnectbuilds.py
Comment (wedamija) [inline src/sentry/models/appconnectbuilds.py:24]: It's slightly risky to add a db level constraint to `Project`, so generally I recommend using `db_constraint=False` here.

File: src/sentry/models/appconnectbuilds.py
Comment (wedamija) [inline src/sentry/models/appconnectbuilds.py:12]: It's probably worth adding this to https://github.com/getsentry/sentry/blob/38abee742ce6a746f5687f02014aded786d99344/src/sentry/deletions/defaults/project.py#L17, since I could imagine we might have a lot of these?

**Comment by flub** [comment]
@tonyo 
> * Are we going to run any other tasks on the new task queue?

I expect over time there will be more tasks added, most imminently there will be one that will periodically check the iTunes session and send and email to the customer if they need to refresh it.

> * What's the expected traffic for the new queue? Like, will it be rising gradually/on-demand, or e.g. we're going to spawn a task for every existing project a few times a day?

We will be spawning a task for every customer that use this feature several times a day.  Initially we'll have few people adopting this and maybe even discourage larger orgs to start using it.  As load goes up we will need to monitor and start optimising the task to do fewer things (fewer App Store Connect API calls).  We may have to adjust new projects onboarding and batch them into smaller tasks to not head-of-line block everyone when a new project starts using the feature, or maybe even create a separate "batch" queue for that if we don't manage otherwise.

Likewise maybe having the same queue for email notifications as for the download tasks will turn out to be a bad idea and we may need to decide to split off the email queue.  I think most of these decisions we'll have to make from monitoring behaviour during onboarding of customers and during EA.

**Comment by flub** [comment]
> This popped into my head when we did the planning: Do we have pagination for these things? Or do we get **all** the releases all the time? Can we add pagination, and maybe take advantage of the sort order (does it have one?) to exit early?

Yes, this has been discussed at various points, we need to monitor this as we onboard people and we may need to start using the API more cleverly to avoid rate-limits etc.

File: src/sentry/models/debugfile.py
Comment (flub) [reply src/sentry/models/debugfile.py:403]: oops, yes.  also no point in formatting debug-messages on prod.

File: src/sentry/lang/native/appconnect.py
Comment (flub) [reply src/sentry/lang/native/appconnect.py:166]: I've been wondering if I should be turning it into a dataclass as well really.  maybe i should bite the bullet and do that

File: src/sentry/models/appconnectbuilds.py
Comment (flub) [reply src/sentry/models/appconnectbuilds.py:12]: makes sense, thanks!

File: src/sentry/models/appconnectbuilds.py
Comment (flub) [reply src/sentry/models/appconnectbuilds.py]: done

**Comment by flub** [comment]
@wedamija @markstory PTAL

File: src/sentry/lang/native/appconnect.py
Comment (untitaker) [reply src/sentry/lang/native/appconnect.py:166]: I mean `Config = Dict[str, Any]` would already help a lot for understanding

File: src/sentry/tasks/app_store_connect.py
Comment (tonyo) [inline src/sentry/tasks/app_store_connect.py:22]: To make this explicit and consistent with other tasks, let's add the `queue` attribute here.

File: src/sentry/tasks/app_store_connect.py
Comment (tonyo) [reply src/sentry/tasks/app_store_connect.py:47]: Let's remove these.

File: src/sentry/tasks/app_store_connect.py
Comment (tonyo) [inline src/sentry/tasks/app_store_connect.py:56]: _remove me_ 🙏

**Review by tonyo — CHANGES_REQUESTED** [review]

File: src/sentry/api/endpoints/project_app_store_connect_credentials.py
Comment (untitaker) [inline src/sentry/api/endpoints/project_app_store_connect_credentials.py:260]: fwiw this now has to be pickle-safe, so you may never move the class definition around or in-flight tasks break

would do the from_json within the task

Comment (tonyo) [reply src/sentry/api/endpoints/project_app_store_connect_credentials.py:260]: Speaking of `config`, how big is that object?

File: src/sentry/tasks/app_store_connect.py
Comment (flub) [reply src/sentry/tasks/app_store_connect.py]: ok, added this back.  Though it is redundant with the `routing_key` arg in `server.py` config.

File: src/sentry/api/endpoints/project_app_store_connect_credentials.py
Comment (flub) [reply src/sentry/api/endpoints/project_app_store_connect_credentials.py]: @untitaker oh, thanks for the hint!  changed that back to be plain json-as-dict

@tonyo my test case is 1924 bytes.  Variation on this will be small, so expect all to be around the 2000 bytes mark.

Comment (flub) [reply src/sentry/api/endpoints/project_app_store_connect_credentials.py]: @tonyo this is kind of a work-around around some (now it seems poor) decisions on how to share responsibilities between the UI and the backend.  If you think this is a problem we can completely remove the argument and only send over the two IDs, only cost is that it'll also need some changes to the UI which i was trying to avoid

**Review by untitaker — APPROVED** [review]
typing looks ok

File: src/sentry/api/endpoints/project_app_store_connect_credentials.py
Comment (tonyo) [inline src/sentry/api/endpoints/project_app_store_connect_credentials.py:346]: As discussed in DMs, it is not recommended to have generic config objects (of variable size) as task arguments, since it might accidentally result in passing KB/MB of data through RabbitMQ. Ideally, we just pass IDs, and then fetch/generate the actual configuration on the consumer side.

Can be fixed in a follow-up PR.

**Review by tonyo — APPROVED** [review]

**Comment by wedamija** [comment]
Migration LGTM, although you'll need to fix the migration conflict. Generally I'd recommend you split up your migration into a separate pr, just so you can get it merged sooner.

File: src/sentry/models/appconnectbuilds.py
Comment (wedamija) [inline src/sentry/models/appconnectbuilds.py:22]: This is probably `core = False`. Core is badly named, but it's just used for import/exporting data in on-prem

Comment (flub) [reply src/sentry/models/appconnectbuilds.py:22]: Could you expand on this?  I basically went with the related table is `__core__ = True` so copied that.  If we set this to false the data will not be exported and after an import this table would be empty?

This would work for here, the next task would re-populate the table and re-download the dSYMs.  In fact if the DIFs are not copied along with an export (they're in filestore).  Actually looking at the `FileBlob` table that is `__core__ = False` so maybe if I understand correctly this forces us to be `__core__ = False` as well.

If you could let me know if my understanding is correct that'd be great.  I'll also split the migration off to its own PR now.

**Comment by flub** [comment]
rebased this onto https://github.com/getsentry/sentry/pull/26864 which is the db migration split off into its own PR

File: src/sentry/models/appconnectbuilds.py
Comment (wedamija) [reply src/sentry/models/appconnectbuilds.py]: Yeah, basically it means that if someone is using this in on-prem, and they want to migrate to Saas, then any rows in this table will not be included in the migration over.

Basically, anything related to config should be `__core__ = True`. So orgs/projects/members/settings etc. Anything related to events/groups/files/releases shouldn't, since they can end up quite large. In practice, since `__core__` is named unclearly, people have applied it kind of haphazardly, and it probably needs an audit at some point.

Comment (flub) [reply src/sentry/models/appconnectbuilds.py]: As a very general comment, I did briefly look for a guide on develop.sentry.dev about adding new tables and creating migrations but didn't find one.  There's a bunch of things I learned stumbling through this that could be made a smoother experience with such a guide 😉

File: src/sentry/api/endpoints/project_app_store_connect_credentials.py
Comment (flub) [reply src/sentry/api/endpoints/project_app_store_connect_credentials.py]: this has been fixed now

**Pull request merged**
